// core/tag_management/tag_crud.js
/**
 * This service provides CRUD functionality for tags
 * @author kvermeer
 */

// Dependencies
var Tag = require('./models/tag.js');

function TagCRUDService() {}

/**
 * Creates a tag
 * @param {string} tagName - the actual text, or name, of the tag
 * @param {int} userId - the user to which this tag belongs
 * @return {void}
 */
TagCRUDService.prototype.createTag = function(tagName, userId) {

    return Tag.create({
        tag_name: tagName,
        user_id: userId
    }).then(function(tag) {
        return {
            tagId: tag.get('tag_id')
        };
    });
}

/**
 * Retrieves a tag
 * @param {int} tagId - the id of the tag to retrieve
 * @return {Promise} -  eventually resolves with a tag
 */
TagCRUDService.prototype.retrieveTag = function(tagId) {

    return Tag.find({
        attributes: ['tag_id', 'user_id', 'tag_name', 'created_at', 'updated_at'],
        where: {
            'tag_id': tagId
        }
    }).then(function(tag) {
        return tag;
    });
}

module.exports = TagCRUDService;
