// core/nugget_management/crud_nugget.js
/**
 * This service provides CRUD functionality for nuggets
 * @author kvermeer
 */

// Dependencies
var Nugget = require('./models/nugget.js');

function NuggetCRUDService() {}

/**
 * Creates a nugget
 * @param {string} content - the content of the nugget
 * @param {string} contentMarkupFormat - the formatting language used for this content
 * @param {int} userId - the user to which this content belongs
 * @param {function} callback - the function to be called after the nugget creation is finished
 *                              this function should take two parameters, the object's id
 *                              and an error
 * @return {void}
 */
NuggetCRUDService.prototype.createNugget = function(content, contentMarkupFormat, userId) {

    return Nugget.create({
        content: content,
        content_markup_format: contentMarkupFormat,
        user_id: userId
    }).then(function(nugget) {
        return {
            nuggetId: nugget.get('nugget_id')
        };
    });
}

/**
 * Retrieves a nugget
 * @param {int} nuggetId - the id of the nugget to retrieve
 * @return {Promise}
 */
NuggetCRUDService.prototype.retrieveNugget = function(nuggetId) {

    return Nugget.find({
        attributes: ['nugget_id', 'user_id', 'content_markup_format', 'content', 'created_at', 'updated_at'],
        where: {
            'nugget_id': nuggetId
        }
    }).then(function(nugget) {
        var flatNugget = {
            nugget_id: nugget.get('nugget_id'),
            user_id: nugget.get('user_id'),
            content_markup_format: nugget.get('content_markup_format'),
            content: nugget.get('content'),
            created_at: nugget.get('created_at'),
            updated_at: nugget.get('updated_at')
        };
        return flatNugget;
    });
}

module.exports = NuggetCRUDService;
