// core/nugget_management/crud_nugget.js
/**
 * This service provides CRUD functionality for nuggets
 * @author kvermeer
 */

// Dependencies
var pg = require('pg');
var Nugget = require('./models/nugget.js');

function NuggetCRUDService() {}

/**
 * Creates a nugget
 * @param {string} content - the content of the nugget
 * @param {string} contentMarkupFormat - the formatting language used for this content
 * @param {int} userId - the user to which this content belongs
 * @param {function} callback - the function to be called after the nugget creation is finished
 *                              this function should take two parameters, the object's id
 *                              and an error
 * @return {void}
 */
NuggetCRUDService.prototype.createNugget = function(content, contentMarkupFormat, userId, callback) {

    if (!content) { // Make sure we are actually recieving something
        callback(null, 'no_content_provided');
    }

    pg.connect(process.env.DATABASE_URL, function(err, client, done) {
        if (err) {
            console.error(err);
        }
        var queryStatement = 'INSERT INTO nugget (content, content_markup_format, user_id) VALUES ($1, $2, $3) RETURNING nugget_id;';
        var values = [content, contentMarkupFormat, userId];
        client.query(queryStatement, values, function(err, result) {
             done();
             if (result && result.rowCount === 1) {
                 var row = result.rows[0];
                 callback(row.nugget_id, null);
             } else {
                 console.error(err);
                 callback(null, 'error_inserting_record');
             }
         });
    });
}

/**
 * Retrieves a nugget
 * @param {int} nuggetId - the id of the nugget to retrieve
 * @param {function} callback - the function to callback when the query has finished
 *                              this function should take two parameters, the object
 *                              and an error
 */
NuggetCRUDService.prototype.retrieveNugget = function(nuggetId, callback) {

    if (!nuggetId) { // Make sure there is a non-null id with some content
        callback(null, 'nugget_id_not_provided');
    }

    Nugget.find({
        attributes: ['nugget_id', 'user_id', 'content_markup_format', 'content', 'created_date'],
        where: {
            'nugget_id': nuggetId
        }
    }).then(function(nugget) {
        console.log(nugget);
        var flatNugget = {
            nugget_id: nugget.get('nugget_id'),
            user_id: nugget.get('user_id'),
            content_markup_format: nugget.get('content_markup_format'),
            content: nugget.get('content'),
            created_date: nugget.get('created_date')
        };
        callback(flatNugget);
    });
}



module.exports = NuggetCRUDService;
