// core/nugget_management/nugget_finder.js
/**
 * A service for finding nuggets based on certain groupings
 * @author kvermeer
 */

// Dependecies
var Nugget = require('./models/nugget.js');

function NuggetFinderService() {}

/**
 * @param {int} userId - the user id of the owner of the nuggers we are gathering
 * @return {Promise} - A promise that will resolve with an array of nugget data
 */
NuggetFinderService.prototype.findNuggetsForUser = function(userId) {

    return Nugget.findAll({
        where: {
            user_id: userId
        }
    }).then(function(nuggets) {
        var flatNuggetArray = [];
        var numberOfNuggetsFound = nuggets.length;
        for (var i = 0; i < numberOfNuggetsFound; i++) {
            var nugget = nuggets[i];
            var flatNugget = {
                nugget_id: nugget.get('nugget_id'),
                user_id: nugget.get('user_id'),
                content_markup_format: nugget.get('content_markup_format'),
                content: nugget.get('content'),
                created_at: nugget.get('created_at'),
                updated_at: nugget.get('updated_at')
            };
            flatNuggetArray.push(flatNugget);
        }
        return flatNuggetArray;
    });
}

module.exports = NuggetFinderService;
