// core/user_management/login_service.js
'use strict';

// Dependencies
var bcrypt = require('bcrypt');
var User = require('./models/user.js');

/**
 * Class containing logic for validating user login
 * @author kvermeer
 */
class LoginService {

    /**
     * Verifies if a user's provided login credentials are correct
     * @param {string} email - email for user
     * @param {string} password - password for logging into user account
     * @return {Promise}
     */
    checkLoginCredentials(email, password) {
        return User.find({
            where: {
                login: email
            }
        }).then(function(user) {
            var userPasswordHash = user.get('password_hash');
            if (bcrypt.compareSync(password, userPasswordHash)) { // Credentials are correct
                return {
                    success: true,
                    userId: user.get('user_id')
                }
            } else {
                return {
                    success: false
                }
            }
        }).catch(function(err) {
            console.log('Unable to find user email: ' + email);
            console.log(err);
            return {
                success: false
            };
        });
    }
}

module.exports = LoginService;
