// core/user_management/user_profile.js
'use strict';

// Dependencies
var User = require('./models/user.js');

/**
 * Service for retrieving user profiles
 * @author kvermeer
 */
class UserProfileService {

    /**
     * Retrieves a user profile
     *
     * @param {int} userId - id of the user to lookup
     * @return {void}
     */
    getUserProfile(userId) {

        return User.find({
            attributes: ['user_id', 'display_name', 'login', 'password_salt', 'password_hash', 'created_at', 'updated_at'],
            where: {
                'user_id': userId
            }
        }).then(function(user) {
            var userProfileData = {
                user_id: user.get('user_id'),
                display_name: user.get('display_name'),
                login: user.get('login'),
                password_salt: user.get('password_salt'),
                password_hash: user.get('password_hash'),
                created_at: user.get('created_at'),
                updated_at: user.get('updated_at')
            }
            return userProfileData;
        });
    }
}

module.exports = UserProfileService;
