// core/user_management/user_profile.js
/**
 * Service for retrieving user profiles
 * @author kvermeer
 */

// Dependencies
var User = require('./models/user.js');

var UserProfileService = function() {}

// Public Methods

/**
 * Retrieves a user profile
 * @param {int} userId - id of the user to lookup
 * @param {function} callback - function to call when we have gather user profile information
 *                              the function should take an object (user profile)
 *                              and an error string as arguments
 * @return {void}
 */
UserProfileService.prototype.getUserProfile = function(userId, callback) {

    User.find({
        attributes: ['user_id', 'display_name', 'login', 'password_salt', 'password_hash', 'created_at', 'updated_at'],
        where: {
            'user_id': userId
        }
    }).then(function(user) {
        var userProfileData = {
            user_id: user.get('user_id'),
            display_name: user.get('display_name'),
            login: user.get('login'),
            password_salt: user.get('password_salt'),
            password_hash: user.get('password_hash'),
            created_at: user.get('created_at'),
            updated_at: user.get('updated_at')
        }
        callback(userProfileData, null)
    });
}

module.exports = UserProfileService;
