// webapp/nugget/controllers/nugget.js
/**
 * Controller for basic nugget operations
 * @author kvermeer
 */
 'use strict';

// Dependencies
var BaseComponents = require('webapp/base');
var NuggetManagement = require('core/nugget_management');
var util = require('util');
var marked = require('marked');

function NuggetController() {}
util.inherits(NuggetController, BaseComponents.BaseController);

/**
 * Method for viewing a pre-existing nugget
 * @param {Express.Request} request - the request object
 * @param {Express.Response} response - the response object
 * @return {void}
 */
NuggetController.prototype.viewNugget = function(request, response) {
    var nuggetId = request.params.nuggetId;

    // Use NuggetCreationServie to actually query the object
    var nuggetCRUDService = new NuggetManagement.NuggetCRUDService();
    nuggetCRUDService.retrieveNugget(nuggetId).then(function(nugget) {
        if (nugget) {
            nugget.content = marked(nugget.content);
            response.render('pages/nugget/view_nugget', nugget);
        } else {
            response.status(500).end();
        }
    });
}

/**
 * Method for serving the page to create a new nugget
 * @param {Express.Request} request - the request object
 * @param {Express.Response} response - the response object
 * @return {void}
 */
NuggetController.prototype.createNuggetPage = function(request, response) {
    this.requireLoginIn(request, response);
    response.render('pages/nugget/create_nugget');
};

/**
 * Method for handling the submit of a new nugget
 * @param {Express.Request} request - the request object
 * @param {Express.Response} response - the response object
 * @return {void}
 */
NuggetController.prototype.createNuggetSubmit = function(request, response) {
    this.requireLoginIn(request, response);
    var content = request.body.nugget_content;

    // Use NuggetCreationServie to actually create the object
    var nuggetCRUDService = new NuggetManagement.NuggetCRUDService();
    nuggetCRUDService.createNugget(content, 'markdown', this.getCurrentUserId(request)).then(function(data) {
        handleCreateNuggetResponse(data, response);
    });
};

function handleCreateNuggetResponse(data, response) {
    response.setHeader('Content-Type', 'application/json');
    var responseData = JSON.stringify({
        data: {
            nugget_id: data.nuggetId
        }
    });
    response.send(responseData);
};

module.exports = NuggetController;
